tokio/src/task/unconstrained.rs

```diff
@@ -34,6 +34,31 @@ where
     }
 }
 
+impl<F: Future> Unconstrained<F>
+where
+    F: Future,
+{
+    /// Gets a reference to the wrapped future.
+    pub fn get_ref(&self) -> &F {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the wrapped future.
+    pub fn get_mut(&mut self) -> &mut F {
+        &mut self.inner
+    }
+
+    /// Gets a pinned mutable reference to the wrapped future.
+    pub fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut F> {
+        self.project().inner
+    }
```

Comment:
```suggestion
```

You can always create a `Pin` later when needed

Reply:
No. `Pin::new()` requires `T` impl `Unpin`.

Reply:
That shouldn't be an issue.

Reply:
Consider this:

```rust
struct State(PhantomPinned);

struct MyFuture {
  some_state: State
}

impl MyFuture {
  fn ready_to_poll(&self) -> bool {
    // check the condition
  }
  fn setup_work(self: Pin<&mut Self>, cx: &mut Context<'_>) {
    // do some setup
  }
}

impl Future for MyFuture {
  type Output = SomeOutput;
  fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
    // poll
  }
}

pin_project! {
  struct MyWrapper {
    #[pin]
    inner: Unconstrained<MyFuture>
  }
}

impl Future for MyWrapper {
  type Output = SomeOutput;
  fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
    if self.inner.get_ref().ready_to_poll() {
      self.project().inner.poll(cx)
    } else {
      self.project().inner.get_pin_mut().setup_work(cx);
      Poll::Pending
    }
  }
}
```

Existing APIs like `tokio::io::BufReader` also have `get_pin_mut()`